File: uva_572.py
MAP=[[0]*32 for _ in range(32)]
q=[]
def char2int(c):
    if c=='@':
        return 1
    else:
        return 0
def bfs(i, j):
    global MAP, q
    MAP[i][j]=0
    while len(q):
        i, j = q[0][0], q[0][1]
        if MAP[i-1][j-1]:
           MAP[i-1][j-1]=0;
           data=[i-1, j-1]
           q.append(data)

        if MAP[i-1][j]:
            MAP[i-1][j]=0;
            data=[i-1, j]
            q.append(data)

        if MAP[i-1][j+1]:
            MAP[i-1][j+1]=0;
            data=[i-1, j+1]
            q.append(data)

        if MAP[i][j-1]:
            MAP[i][j-1]=0;
            data=[i, j-1]
            q.append(data)

        if MAP[i][j+1]:
            MAP[i][j+1]=0;
            data=[i, j+1]
            q.append(data)

        if MAP[i+1][j-1]:
            MAP[i+1][j-1]=0;
            data=[i+1, j-1]
            q.append(data)

        if MAP[i+1][j]:
            MAP[i+1][j]=0;
            data=[i+1, j]
            q.append(data)

        if MAP[i+1][j+1]:
            MAP[i+1][j+1]=0;
            data=[i+1, j+1]
            q.append(data)

        q.pop(0)
while True:
    try:
        r, c=input().split()
        r, c, = int(r), int(c)
    except:
        break
    if not r and not c:
        break
    n=0
    MAP=[[0]*102 for _ in range(102)]
    for i in range(1, r+1):
        data=input()
        for j in range(1, c+1):
            MAP[i][j]=char2int(data[j-1])

    for i in range(1, r+1):
        for j in range(1, c+1):
            if MAP[i][j]:
                n+=1
                start=[i, j]
                q.append(start)
                bfs(i, j)
    print(n)

File: uva_572.cpp
# include <queue>
# include <iostream>
using namespace std;
int MAP[102][102];
queue<int*> q;

int char2int(char c){
    if (c=='@')
        return 1;
    else
        return 0;
}
void bfs(int i, int j){
    MAP[i][j]=0;
    while (!q.empty()){
        i=q.front()[0], j=q.front()[1];
        q.pop();
        if(MAP[i-1][j-1]){
           MAP[i-1][j-1]=0;
           int *data=new int[2];
           data[0]=i-1, data[1]=j-1;
           q.push(data);
        }
        if(MAP[i-1][j]){
            MAP[i-1][j]=0;
            int *data=new int[2];
            data[0]=i-1, data[1]=j;
            q.push(data);
        }
        if(MAP[i-1][j+1]){
            MAP[i-1][j+1]=0;
            int *data=new int[2];
            data[0]=i-1, data[1]=j+1;
            q.push(data);
        }
        if(MAP[i][j-1]){
            MAP[i][j-1]=0;
            int *data=new int[2];
            data[0]=i, data[1]=j-1;
            q.push(data);
        }
        if(MAP[i][j+1]){
            MAP[i][j+1]=0;
            int *data=new int[2];
            data[0]=i, data[1]=j+1;
            q.push(data);
        }
        if(MAP[i+1][j-1]){
            MAP[i+1][j-1]=0;
            int *data=new int[2];
            data[0]=i+1, data[1]=j-1;
            q.push(data);
        }
        if(MAP[i+1][j]){
            MAP[i+1][j]=0;
            int *data=new int[2];
            data[0]=i+1, data[1]=j;
            q.push(data);
        }
        if(MAP[i+1][j+1]){
            MAP[i+1][j+1]=0;
            int *data=new int[2];
            data[0]=i+1, data[1]=j+1;
            q.push(data);
        }
    }
}

int main(void){
    int r, c, n;
    char C;
    while(cin >>r>>c){
        if (!r && !c)
            break;
        n=0;
        for (int i=0; i<102; ++i)
            for (int j=0; j<102; ++j)
                MAP[i][j]=0;

        for (int i=1; i<=r; ++i){
            for (int j=1; j<=c; ++j){
                cin >> C;
                MAP[i][j]=char2int(C);
            }
        }

        for (int i=1; i<=r; ++i)
            for (int j=1; j<=c; ++j)
                if(MAP[i][j]){
                    ++n;
                    int data[2]={i, j};
                    q.push(data);
                    bfs(i, j);
                }
        cout << n << endl;

    }
    return 0;
}